[HigherOrderOp] track input mutations via tensor version

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115451

Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from @zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. Also modify existing tests that aliasing inputs.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

ydwu4 committed:
[HigherOrderOp] track input mutations via tensor version

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning buffer mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, specuate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/output_graph.py:
```diff
@@ -1829,6 +1842,19 @@ def maybe_lift_tracked_freevar_to_input(self, arg):
         return self.lift_tracked_freevar_to_input(arg)
 
+    def track_mutations_for_tensor_inputs(self, proxies: List[fx.Proxy]):
+        self.tensor_input_versions.update(
```
The version counter logic gets disabled when torch.inference_mode is active. I don't think this gets used often, but I could be wrong. Could we:
- not track this information when torch.inference_mode is active
- raise unimplemented in cond if torch.inference_mode is active

zou3519 replied:
Actually, I'm not sure that torch.inference_mode isn't used often.

zou3519 replied:
Proposal #2, that doesn't use version counters:
- once we have captured a graph, we run clones of the input FakeTensors through the graph with a torch_dispatch mode active
- the torch_dispatch mode will raise an error if any inputs are being mutated. We can tell this by looking at the schema of the operators.

Downsides: doesn't really work with subclasses

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 commented:
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52096595).

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. Also modify existing tests that aliasing inputs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -116,6 +116,41 @@ def _call_function_and_unflatten_output(tx, fn, args, kwargs, ret_vt, ret_treesp
 
+# NOTE:
+# Mutation detection: We detect mutations by checking the example_value's _version
+# This approach has a corner case that if torch.compile is wrapped
+# by an inference_mode(True) context manager, the tensor version will not be upated.
```
Can we add some tests that cond still errors out in a situation where it should if it is under torch.inference_mode?

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -661,30 +696,63 @@ def _insert_or_replace_phs(new_args, name_suffix):
+                mutated_inputs, aliased_inputs = _detect_input_mutations_and_aliasing(
+                    tx, module, combined_args
+                )
+            except Exception as e:
+                unimplemented(
```
try-except with bare Exception is dangerous, do we need this?

ydwu4 replied:
Yeah, we can be more specific about the exception.

zou3519 reviewed (commented):
Code looks good to me, but we should run a test case under inference_mode for completeness

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. Also modify existing tests that aliasing inputs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. Also modify existing tests that aliasing inputs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. Also modify existing tests that aliasing inputs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] track input mutations via tensor version"


Fixes #115277 by banning tensor buffer, tensor input and tensor lifted clousures mutations and requiring users to rewrite their code for now. It implements the idea of using tensor._version to detect mutations from zou3519. After this PR, speculate_subgraph will return input proxies that gets mutated. Different HOOs can choose different strategies to make use of this info.

We might be able to support these mutations for HOOs but need to wait for user requests and decide.

Test Plan:
Added new tests for cond. Also modify existing tests that aliasing inputs.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D52096595](https://our.internmc.facebook.com/intern/diff/D52096595)

[ghstack-poisoned]

ydwu4 commented:
Tests added!
